crates/randcast-mock-demo/src/main.rs

```diff
@@ -0,0 +1,239 @@
+            bincode::serialize(&public_poly.eval(i as u32).value).unwrap(),
+            vec![],
+        );
+        println!("{}-res: {}", i, res);
+    });
+
+    let group = controller.get_group(1);
+
+    println!("group state: {}", group.state);
+
+    println!("group committers: {:?}", group.committers);
+
+    // let msg = rand::random::<[u8; 32]>().to_vec();
+
+    let msg = String::from("ujehwsndfgljkhrlkg");
+
+    println!("An user is requesting a randomness... msg seed: {}", msg);
+
+    let request_res = controller.request(msg.clone());
+
+    println!("request_res: {}", request_res);
+
+    println!("A signature task is emitting...");
+
+    let signature_task = controller.emit_signature_task();
+
+    let signature_index = signature_task.index;
+
+    // generates a partial sig with each share from the dkg
+    let partial_sigs = outputs
+        .iter()
+        .map(|output| G1Scheme::<BLS12_381>::partial_sign(&output.share, &msg.as_bytes()).unwrap())
+        .collect::<Vec<_>>();
+
+    // committer verify the partial threshold signatures first
+    partial_sigs.iter().for_each(|partial_sig| {
+        G1Scheme::<BLS12_381>::partial_verify(&public_poly, &msg.as_bytes(), &partial_sig).unwrap();
+    });
+
+    // then aggregates them
+    let sig = G1Scheme::<BLS12_381>::aggregate(t, &partial_sigs).unwrap();
+
+    // committer verify the threshold signature first
+    G1Scheme::<BLS12_381>::verify(&pubkey, &msg.as_bytes(), &sig).unwrap();
+
+    println!("Committers are committing result of the signature task...");
+
+    (0..n).for_each(|i| {
+        // the participant list to be rewarded by this signature task
+        let mut partial_signatures: HashMap<String, Vec<u8>> = HashMap::new();
+
+        partial_sigs
+            .iter()
+            .enumerate()
+            .for_each(|(i, partial_sig)| {
+                partial_signatures.insert(String::from("0x") + &i.to_string(), partial_sig.clone());
+            });
+
+        let res = controller.fulfill(
+            String::from("0x") + &i.to_string(),
+            signature_index,
+            sig.clone(),
+            partial_signatures,
+        );
+
+        println!("{}-res: {}", i, res);
+    });
+
+    let randomness_output = controller.get_last_output();
+
+    println!("randomness output: {}", randomness_output);
+
+    println!("finish.");
+
+    Ok(())
+}
+
+async fn run_dkg<C, S>(
+    board: &mut InMemoryBoard<C>,
+    phase0s: Vec<joint_feldman::DKG<C>>,
+    rng: &mut ThreadRng,
+) -> Vec<DKGOutput<C>>
+where
+    C: Curve,
+    // We need to bind the Curve's Point and Scalars to the Scheme
+    S: Scheme<Public = <C as Curve>::Point, Private = <C as Curve>::Scalar>,
+{
+    // Phase 1: Publishes shares
+    let mut phase1s = Vec::new();
+    for phase0 in phase0s {
+        phase1s.push(phase0.run(board, rng).await.unwrap());
+    }
+
+    // Get the shares from the board
+    let shares = board.shares.clone();
```

Comment:
Just use the reference, maybe don't need to clone it.

Reply:
I see and this seems to be a typical example of rust restriction on ownership from my limited experience. That is, `shares` is a field of `board`, at line 166 an immutable borrow of `board` occurs here(if without `clone()`), at line 171 a mutable borrow of `board` will be needed for `run(self, board: &mut B, arg: T)` to go. Although we know during `run` fn `shares` from `&mut board` will not be updated, but compiler forbid it at the beginning without any exception. What should we do to make the code more "rusty"? :D

Reply:
why not just pass `&board` to `parse1.run`?

Reply:
It's `phase.run` actually, representing an abstraction of different phases process. The first parameter `board` here is declared with `#[async_trait(?Send)]` to write messages so that `&mut` as reference is necessary, and the second parameter `arg` is the input in this phase to handle and then send to the board. In a real environment `arg` is not directly taken from `board` but from a `eth_call`.